plugins/image-prioritizer/class-image-prioritizer-video-tag-visitor.php

```diff
@@ -43,7 +62,42 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 
 		$xpath = $processor->get_xpath();
 
-		// TODO: If $context->url_metric_group_collection->get_element_max_intersection_ratio( $xpath ) is 0.0, then the video is not in any initial viewport and the VIDEO tag could get the preload=none attribute added.
+		$max_element_width = 0;
+
+		$denormalized_elements = $context->url_metric_group_collection->get_all_denormalized_elements()[ $xpath ] ?? array();
+
+		foreach ( $denormalized_elements as list( , , $element ) ) {
+			$max_element_width = max( $max_element_width, $element['boundingClientRect']['width'] ?? 0 );
+		}
+
+		$poster_id = attachment_url_to_postid( $poster );
+
+		if ( $poster_id > 0 && $max_element_width > 0 ) {
+			$smaller_image_url = wp_get_attachment_image_url( $poster_id, array( (int) $max_element_width, 0 ) );
+			$processor->set_attribute( 'poster', $smaller_image_url );
+		}
```

Comment:
One possible consideration to make here is that if there are no URL Metrics gathered for the largest breakpoint group that in this case the poster should not be reduced in size. The reason being is that if there are only mobile visitors from whom URL Metrics have been gathered, as soon as a desktop visitor visits they could see a poster that is much lower-resolution than it should be.

Maybe this is an edge case that we shouldn't worry about right now, but it's something to keep in mind.

Reply:
Great point, I think this is relevant to consider. I'd say we should be cautious in such a case and prioritize a "not broken" experience over better performance. Basically, unless we have data for all three viewports, we cannot reasonably make a decision to optimize.

Reply:
Related to this is https://github.com/WordPress/performance/pull/1596/files#r1803569058 where I realize that lazy-loading may currently be incorrectly applied to an element on desktop if it is not visible on mobile and only URL Metrics for mobile have been collected. But if a site only ever gets mobile traffic, should the lack of desktop data be a blocker?

Reply:
I think a low-res poster image on desktop if there's a lack of desktop traffic is acceptable.

But we can be more strict for now and then see how well it works. So how can I easily check whether values exist for every viewport? :-)

Reply:
I believe this would do the trick:

```php
$widest_group = array_reduce(
	iterator_to_array( $context->url_metric_group_collection ),
	static function ( $carry, OD_URL_Metric_Group $group ) {
		return ( null === $carry || $group->get_minimum_viewport_width() > $carry->get_minimum_viewport_width() ) ? $group : $carry;
	}
);
foreach ( $widest_group as $url_metric ) {
	foreach ( $url_metric->get_elements() as $element ) {
		if ( $element['xpath'] === $xpath ) {
			$max_element_width = max( $max_element_width, $element['boundingClientRect']['width'] ?? 0 );
			break; // Move on to the next URL Metric.
		}
	}
}
```

Reply:
In reality, the last item in `iterator_to_array( $context->url_metric_group_collection )` will have the widest viewport since `OD_URL_Metric_Group_Collection` sorts the breakpoints before constructing the groups.

Reply:
It would probably make sense to add a helper to `OD_URL_Metric_Group_Collection` which returns the widest group (i.e. desktop). But we can do that later.

Reply:
OK I can give that a try tomorrow.

Reply:
I didn't want to lead you astray, so I went ahead and tested it out in 3521f9ec2312477e6813b0c4975c5e3f6b40819c. It seems to work! There are two test cases, one for when desktop URL metrics are collected and another when they are absent.

Reply:
Amazing, thanks a lot!